[MSE] MediaSource playback rapidly loops after changing audio track

jyavenard opened the pull request:
#### 4aa58c3908a0aa119de1ca6c1712154db4c6d515
<pre>
[MSE] MediaSource playback rapidly loops after changing audio track
<a href="https://bugs.webkit.org/show_bug.cgi?id=276421">https://bugs.webkit.org/show_bug.cgi?id=276421</a>
<a href="https://rdar.apple.com/131577794">rdar://131577794</a>

Reviewed by Youenn Fablet.

The MediaPlayerPrivateMediaSourceAVFObjC wait for all audio and video renderers to
contain samples before changing the AVSampleBufferRenderSynchronizer&apos;s rate or
set it to 0 if one was missing data. This is used to force playback to stall
if we are missing decoded data.

We should normally only have one AVSampleBufferAudioRenderer when there is only
one active audio track.
This AVSampleBufferAudioRenderers is stored in a hash map using the TrackID as key.
When a new init segment is processed that changed the TrackID rather than re-use
the existing AVSampleBufferAudioRenderer we would end up creating a new one.
This would cause an infinite stall due to the condition described above as
we are waiting for all audio renderers to have data.
When the TrackID change mid-flight, the SourceBuffer notifies the SourceBufferPrivate
via a call to `updateTrackIds`. We create a local override to update the map
of audio renderers with the new trackID so that we don&apos;t incorrectly create a new one
and re-use the existing one.

Added test: we add data from two audio tracks with different ID and make sure that the video&apos;s currentTime
progresses.

* LayoutTests/media/media-source/content/test-48kHz-new-track-id.m4a: File was invalid, only init segment was correct
* LayoutTests/media/media-source/content/test-48khz-new-track-id-manifest.json: Update media segment offsets which were incorrect.
* LayoutTests/media/media-source/media-source-audio-track-id-switch-play-expected.txt: Added.
* LayoutTests/media/media-source/media-source-audio-track-id-switch-play.html: Added.
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h: Add audioRendererForTrackID convenience method.
When we only have one audio track, we should ignore the TrackID when looking for the audio renderer as it can change over time.
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::isMediaSampleAllowed const):
(WebCore::SourceBufferPrivateAVFObjC::updateTrackIds):
(WebCore::SourceBufferPrivateAVFObjC::appendInternal):
(WebCore::SourceBufferPrivateAVFObjC::flush):
(WebCore::SourceBufferPrivateAVFObjC::enqueueSample):
(WebCore::SourceBufferPrivateAVFObjC::isReadyForMoreSamples):
(WebCore::SourceBufferPrivateAVFObjC::didBecomeReadyForMoreSamples):
(WebCore::SourceBufferPrivateAVFObjC::notifyClientWhenReadyForMoreSamples):

Canonical link: <a href="https://commits.webkit.org/283049@main">https://commits.webkit.org/283049@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c78f665185d6459947c38e9fc0ef8ec184b12843

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65038 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44405 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17652 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69062 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15644 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67156 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52188 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15926 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52251 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10804 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68104 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41056 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56291 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32873 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37725 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13662 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14520 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59637 "1 api test failed or timed out") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14000 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70767 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8990 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13480 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59578 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9022 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56351 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59819 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/list-markers, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/selection/menulist (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14349 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7419 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1097 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40217 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41294 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42475 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41038 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
```diff
@@ -350,13 +348,11 @@ ALLOW_NEW_API_WITHOUT_GUARDS_END
-    };
 ALLOW_NEW_API_WITHOUT_GUARDS_BEGIN
-    HashMap<RetainPtr<CFTypeRef>, AudioRendererProperties> m_sampleBufferAudioRendererMap;
+    Vector<RetainPtr<AVSampleBufferAudioRenderer>> m_sampleBufferAudioRenderers;
```
`HashSet<RetainPtr<AVSampleBufferAudioRenderer>>` is probably what would make most sense.

jyavenard committed:
[MSE] MediaSource playback rapidly loops after changing audio track
https://bugs.webkit.org/show_bug.cgi?id=276421
rdar://131577794

Reviewed by Youenn Fablet.

The MediaPlayerPrivateMediaSourceAVFObjC wait for all audio and video renderers to
contain samples before changing the AVSampleBufferRenderSynchronizer's rate or
set it to 0 if one was missing data. This is used to force playback to stall
if we are missing decoded data.

We should normally only have one AVSampleBufferAudioRenderer when there is only
one active audio track.
This AVSampleBufferAudioRenderers is stored in a hash map using the TrackID as key.
When a new init segment is processed that changed the TrackID rather than re-use
the existing AVSampleBufferAudioRenderer we would end up creating a new one.
This would cause an infinite stall due to the condition described above as
we are waiting for all audio renderers to have data.
When the TrackID change mid-flight, the SourceBuffer notifies the SourceBufferPrivate
via a call to `updateTrackIds`. We create a local override to update the map
of audio renderers with the new trackID so that we don't incorrectly create a new one
and re-use the existing one.

Added test: we add data from two audio tracks with different ID and make sure that the video's currentTime
progresses.

* LayoutTests/media/media-source/content/test-48kHz-new-track-id.m4a: File was invalid, only init segment was correct
* LayoutTests/media/media-source/content/test-48khz-new-track-id-manifest.json: Update media segment offsets which were incorrect.
* LayoutTests/media/media-source/media-source-audio-track-id-switch-play-expected.txt: Added.
* LayoutTests/media/media-source/media-source-audio-track-id-switch-play.html: Added.
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h: Add audioRendererForTrackID convenience method.
When we only have one audio track, we should ignore the TrackID when looking for the audio renderer as it can change over time.
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::isMediaSampleAllowed const):
(WebCore::SourceBufferPrivateAVFObjC::updateTrackIds):
(WebCore::SourceBufferPrivateAVFObjC::appendInternal):
(WebCore::SourceBufferPrivateAVFObjC::flush):
(WebCore::SourceBufferPrivateAVFObjC::enqueueSample):
(WebCore::SourceBufferPrivateAVFObjC::isReadyForMoreSamples):
(WebCore::SourceBufferPrivateAVFObjC::didBecomeReadyForMoreSamples):
(WebCore::SourceBufferPrivateAVFObjC::notifyClientWhenReadyForMoreSamples):

Canonical link: https://commits.webkit.org/283049@main
